### ISSUE TITLE
Every property in Typescript defined as required

### ISSUE THREAD

**Issue by edvald** [kubernetes-client/javascript#52]
It seems the conversion from Swagger to TypeScript incorrectly makes every parameter on request body types required. In a lot of cases the request parameters contain all of the properties that you'd expect when fetching objects, which ofter you can't use in the request body, so the only feasible workaround is to cast to with `<any>` or `Partial<T>`. For example, `status` is required in the interface for the body for `createNamespace`, which of course doesn't make sense when creating a namespace.

I'm not sure if this is an upstream issue with `swagger-codegen` or the configuration, but for the library to work as expected, there probably needs to be a separation between types returned from read/list calls, and the types used for create/patch calls.

Let me know if [we](https://github.com/garden-io/garden) can help with this, we're likely to be heavy users of the library so we're happy to contribute, but would probably need some pointers to get started .)

**Comment by JorritPosthuma** [comment]
Hi @edvald, I'm curious to if/how you solved it in the meantime? Right now I'm doing something like this:

```typescript
// First, we define a type that captures a collection of methods we want to redefine
type CreateApiMethod<B, R> = (namespace: string, body: B, pretty?: string) => Promise<R>;
// This creates the new API type. For every method name K of the API, it gets the method T[K].
// If T[K] matches the format of CreateApiMethod, then infer the Body type B, and Return type R,
// and redeclare the method as replacing the Body type with its DeepPartial counterpart.
type PartialApi<T> = { [K in keyof T]: T[K] extends CreateApiMethod<infer B, infer R> ? CreateApiMethod<DeepPartial<B>, R> : T[K] };

const config = new KubeConfig();
config.loadFromDefault();

// Here, of course TypeScript complains that the client type doesn't match the "partiallized" client type. Just ignore it :).
// @ts-ignore
export const apps = config.makeApiClient(Apps_v1Api) as PartialApi<Apps_v1Api>;
```

This snippet only changes the create methods, but you can do the same for the `delete`, `replace`, ... methods.

**Comment by edvald** [comment]
That's actually a much better solution than we'd gone with, which was basically casting values at the call-site. I'd be in favor of applying this, with the addition of similar wraps around `delete` etc., in the `makeApiClient` method.

**Comment by geekflyer** [comment]
just came across this. On this occasion I just wanna mention that https://github.com/pulumi/pulumi-kubernetes also has typescript definitions which are generated (afaik also from the openapi spec) and their types are correct, i.e. only actually required fields are marked requires in the types.

**Comment by flyboarder** [comment]
@geekflyer they are using a go build to generate the files, this project uses the generator script located in https://github.com/kubernetes-client/gen

**Comment by kraenhansen** [comment]
How are people even using this library from TypeScript? I want to create a namespace and the types require me to fill in `clusterName`, `creationTimestamp` and a lot of other obviously optional parameters.

I would love to see a TypeScript example that simply creates a namespace:

```typescript
await this.k8s.createNamespace({
    metadata: {
        name: 'my-awesome-namespace',
    },
});
```

**Comment by silasbw** [comment]
This issue was fixed with https://github.com/kubernetes-client/javascript/pull/268.  The latest versions of the library include the fix.  Should we close this @brendandburns ?

**Comment by drubin** [comment]
/close
Fixed in #268

We can reopen if there are any other issues.
